### ISSUE TITLE
Different kinds of engines and avoiding copies

### ISSUE THREAD

**Issue by germasch** [ornladios/ADIOS2#1315]
So while trying to avoid stirring up specifics on API design, I think it would be useful to have a general overview of what the different engines do under the hood, and where copies / buffering can potentially be avoided.

I certainly haven't looked in detail what the various engines do, so I may be wrong on aspects, and I'll definitely be incomplete. I think this could also be useful for compiling a list of behavior that the app developers need to know, in particular about which calls for which engines might be blocking, or, e.g., MPI collective, and under which circumstances.

### BP3

I think I have a somewhat decent understanding what happens with file-backed BP3, when using one file per proc.

* `PutSync` will take the data and copy it into its internal buffer, which is essentially the content of (or a chunk of, if too large to buffer) the eventual file kept in memory. Should non-blocking. Might trigger a realloc of the buffer to extend it. Might trigger a flush when max buffer size is set and reached.
* `PutDeferred` only stores the data pointer and associated info for later use in `PerformPuts`.
* `PerformPuts` essentially runs `PutSync` on all data pointers previously stored by `PutDeferred`, so the same copy and blocking behavior applies.
* `EndStep` may flush, otherwise doesn't do much beyond `PerformPuts`.

In the typical case, using either `PutSync` or `PutDeferred` will involve two copies to get the data into the file. (1) from the application memory into the BP3 buffer, (2) from BP3 buffer to kernel space. (I'll stop here. if an actual disk is attached, it the controller may DMA from the kernel buffer. For remote filesystems, which I suppose is the typical use case, the network adapter may be doing the DMAing, or RDMAing). The Span interface allows to avoid the copy (1), with some caveats, isn't otherwise used to extend lifetime of the user provided data.

TODO: What happens when MPI aggregation is involved? I suppose MPI blocking or even collective behavior may be involved.

### InsituMPI

(I'm just going by what I see looking through the code)

* `PutSync` is only supported for single values, not for array data. Single values will be buffered locally. If max buffer size is set and reached, it will throw an exception (not handled). Otherwise will not block.
* `PutDeferred` will `MPI_Isend` the data if definitions are locked on both sender and receiver side. Otherwise, the data pointer will be retained for later use in PerformPuts.  As above, may run out of buffer space, which is not handled. Otherwise will not block.
* `PerformPuts` collectively communicates metadata info. Will then `MPI_Isend` `PutDeferred` variables that not already sent in `PutDeferred above. Will not wait for the `MPI_Isend`s to complete.
* `EndStep`  will wait for the `MPI_Isend`s to complete, and wait for acknowledgement from readers (collective). May potentially block a long time if the reader isn't ready for the step yet.

The Span interface, if it were implemented, could allow to avoid a copy into the internal buffer. The way it looks to me, that copy also could be avoided when having `PutDeferred` directly `MPI_Isend()` the user buffer, but in this case `PerformPuts` would have to `MPI_Wait`, instead of waiting until `EndStep`.

### SST

(I'm just going by what I see looking through the code, and I'm only looking at BP3 marshaling)

* `PutSync`: copies into BP3 buffer. Won't block. Doesn't check resizeResult, so either it is impossible to set MaxBufferSize, or otherwise there's a bug in that it at least should throw an exception because it can't handle a flush
* `PutDeferred`: actually just calls `PutSync`, so same behavior
* `PerformPuts`: does nothing (since nothing is ever deferred)
* `EndStep`: provides the timestep to the receiver. Does not wait for the receiver to actually receive, consequently, the internal buffer needs to be retained until a later time.

Copy into BP3 buffer could be avoided using Span interface. Could also potentially be avoided by starting communication in `PutDeferred` and waiting for completion in `PerformPuts`, that, however, may not be desirable as it looks to me that the goal is to not have sender and receiver go in lockstep, ie, prevent blocking in `PerformPuts` or `EndStep`.

### Other engines: TBD

A lot of the current behavior is implied by the use of underlying BP3 marshaling in the various writer engines (SST has an alternative, but I haven't looked a that). Since that is built around a single buffer, it requires a copy to get into the right position, unless the Span interface is used. A side effect of maintaining its own buffer is that it essentially assumes ownership of the user provided data (by copying it). 

If people want to extend / preserve / correct this list, one option would be to put it into a wiki page. Though it looks like people like me then wouldn't be able to edit it ;(

**Comment by williamfgc** [comment]
@germasch this is a good overview, thanks. I had a specs document (including flow diagrams), but no one had the time to maintain/update and fell off in our list of priorities since many of the Engines functionalities you outlined were needed for certain use cases. 

Since the interface is pointer based I make the distinction between `data pointer` (`&i[0]`) readiness, and `data content` (`i[0]`) readiness:

- `Put in sync mode`: `data pointer` and `data content` must be ready

- `Put in deferred mode`: `data pointer` must be ready, `data contents` must be ready at the first encounter of PerformPuts/EndStep/Close. Deferred mode doesn't prevent the user from modifying the `data contents` between Put and PerformPuts/EndStep/Close, but we don't promote this practice as the user might end up modifying/corrupting their own `data pointer`, e.g. resize)

- `Put returning a span`: `data pointer` (provided by adios2) and `data contents` are ready with the option to return the `data pointer` as a `span` to modify `data contents`, `data contents` must be ready at first encounter with PerformPuts/EndStep/Close. In this case it is intended to populate the `data contents`, since `span` by definition doesn't allow the dynamic `data pointer` from `span.data()` to be corrupted by the user 

The same contracts applies to Get, unless we decide to return a span with const/non-const iterators (but that's a major topic that requires a use cases).

**Comment by williamfgc** [comment]
> A side effect of maintaining its own buffer is that it essentially assumes ownership of the user provided data (by copying it).

You're correct, we copy the data, not the data pointer, we own the copy not the original data. Also, buffering is absolutely required for compression.

> (2) from BP3 buffer to kernel space. (I'll stop here...

Yes, POSIX (or fstream/stdio) are considered our "transport" layer. The BP file engine stops there unless new APIs allow more low-level control into the hardware, e.g. I had a `Lustre` branch somewhere.

**Comment by williamfgc** [comment]
> TODO: What happens when MPI aggregation is involved? I suppose MPI blocking or even collective behavior may be involved.

The private behavior of MPI aggregation (substreams to the user) is that it creates a pipeline between ranks of the same aggregators group as only one rank is pointing to a substream, at EndStep/PerformPuts/Close. It can be seen as a hybrid approach or "partially" blocking, since it's "blocking" at the group level, but each group is independent.  Note that I am not including forming the Collective metadata as it's orthogonal

**Comment by pnorbert** [comment]
One correction: InSituMPI never buffers arrays. It will use the original
pointer passed in Put() in the MPI Isend operations.

To support Put in sync mode, this engine would need to do buffering.

**Comment by germasch** [comment]
> One correction: InSituMPI never buffers arrays. It will use the original pointer passed in Put() in the MPI Isend operations. To support Put in sync mode, this engine would need to do buffering.

Thanks, I read the code wrong. I think that's what makes this thread useful, at least to me. I edited my initial post accordingly. However, I did delete also delete the part about "MPI_Isends still being in flight after PerformPuts() is safe, because the data is internally buffered".  Since the data is not buffered, between PerformPuts() and EndStep() the MPI_Isends are still in flight, but at the same time, my understanding is that the app is allowed to modify them?

**Comment by germasch** [comment]
> @germasch this is a good overview, thanks. 

So actually I don't think I organized things particularly well or clear, I guess this might be considered useful for my own education, but I hope in the end it'll contribute to clarifying some issues more generally.

In the interest of trying to keep this thread useful, let's try to not argue about the Span interface here (I will say, though, that the exercise of writing the initial post showed me that there are Engines beyond BP3 can benefit from what Span allows. I actually thought all 3 could, though I did misread the Insitu one, so I don't think it applies there.)

> * `Put in sync mode`: `data pointer` and `data content` must be ready
> * `Put in deferred mode`: `data pointer` must be ready, `data contents` must be ready at the first encounter of PerformPuts/EndStep/Close. Deferred mode doesn't prevent the user from modifying the `data contents` between Put and PerformPuts/EndStep/Close, but we don't promote this practice as the user might end up modifying/corrupting their own `data pointer`, e.g. resize)

So I think you guys need to work out what the semantics are supposed to be. I was under the impression that both @pnorbert and @eisenhauer agreed that both versions of `Put` above indicate data readiness. That's how I understood it right from the start, too, though looking at the docs, they don't seem to actually be specific about it. InsituMPI definitely expects readiness at PutDeferred time, too.
